ReadMe.md
TouchTerrain
============

TouchTerrain converts publically accessible digital elevation data into digital model files (STL or OBJ) 
suitable for 3D printing. It comes as both a standalone program and a server which creates 3D model files and makes
them available for download.

TouchTerrain is developed by Chris Harding and Franek Hasiuk, of Iowa State
University's [GeoFabLab](http://www.public.iastate.edu/~franek/gfl/gfl.html).

To see the server version in action, go to [touchterrain.geol.iastate.edu](http://touchterrain.geol.iastate.edu) 

For more in-depth information read our paper in Computers & Geosciences:
TouchTerrain: A simple web-tool for creating 3D-printable topographic models, Volume 109, December 2017, Pages 25-31, https://doi.org/10.1016/j.cageo.2017.07.005

Getting Started
===============

Unless you're planning to only use the stand-alone version to import a DEM from a rasterfile (see importedDEM under Standalone), you will need to install Google Earth Engine for Python 2. The directions on GEE's [site](https://developers.google.com/earth-engine/python_install) are a good place to start, but additional information is provided in this repository
in the file `standalone/TouchTerrain_standalone_installation.pdf`.

If you do not already have an account with Google Earth Engine, you will need to
apply for one from Google. This process is typically painless. Once a request is
submitted Google usually approves it within about a day.

Various Python modules must be installed, as detailed in
`standalone/TouchTerrain_standalone_installation.pdf`.

Once Earth Engine is installed you can run either
`standalone/TouchTerrain_standalone.py`
or
`server/TouchTerrain_app.py`
from within their respective directories.



Standalone
==========

`TouchTerrain_standalone.py` uses DEM data from Google Earth Engine of from a local raster file (geotiff, etc.) to
create a 3D model suitable for 3D printing. This model, possibly consisting of several files (tiles), is saved in a ZIP along with an info file describing the properties used to produce it. The downsampled and UTM projected DEM for the area is saved as DEM.tif

`TouchTerrain_standalone.py` reads in a JSON configuration file such as the one
at `standalone/example_config.json`. To run it in a python shell, go into the standalone folder and type

`python TouchTerrain_standalone.py example_config.json`

Note: you will have to install an bunch of modules this to actually work. You can use a local DEM file via the `importedDEM` parameter but if you want to use the Google Earth Engine curated DEMs, you need to get a Google Earth Engine license (free). See the TouchTerrain_standalone_installation.pdf for more on that.

The JSON config file has the following format

```
{
"DEM_name": "USGS/NED", 
"basethick": 1, 
"bllat": 44.50185267072875, 
"bllon": -108.25427910156247, 
"fileformat": "STLb", 
"ntilesx": 1, 
"ntilesy": 1, 
"printres": 0.5, 
"tile_centered": true, 
"tilewidth": 80, 
"trlat": 44.69741706507476, 
"trlon": -107.97962089843747, 
"zip_file_name": "terrain", 
"zscale": 1.0,
"CPU_cores_to_use": 0, 
"max_cells_for_memory_only": 1000000,
"importedDEM": null
}
```

The syntax of this file is as follows:

 * `DEM_name`:     (resolutions are approximate and strictly true only at the equator!) 
   * USGS/NED: 10 m, continental USA only
   * USGS/SRTMGL1_003: 30 m, "worldwide" but not very far north
   * USGS/GMTED2010: 90 m, truly worldwide
   * NOAA/NGDC/ETOPO1: 1000 m, worldwide, with bathymetry
 
 * `basethick`:     A layer of material this thick will be added below the 
                    entire model. This is particularly important for models 
                    with long, deep valleys, which can cause the model to shine through or 
                    if the base is not thick enough. A base thickness of at least twice the
                    filament thickness is recommended.

 * `bllat`:         Bottom-left latitude of area to be printed
 * `bllon`:         Bottom-left longitude
 * `trlat`:         Top-right latitude
 * `trlon`:         Top-right longitude

 * `fileformat`: file format for 3D model file. Units are in meters, you will have to scale them down to print.    
   - obj: wavefront obj (ascii)  (no normals, no texture coordinates yets ...)
   - STLa: ascii STL (with normalized normals)
   - STLb: binary STL (with normalized normals), prefered format

 * `ntilesx`:       Divide the x axis evenly among this many tiles. This is
                    useful if the area being printed would be too large to fit
                    in the printer's bed.
 * `ntilesy`:       See `ntilesx`, above.
 
 * `tilewidth`:     The width of a tile in mm, tile height will be calculated from the aspect ratio of your area.

 * `printres`:      A realistic resolution for your printer along the buildplate, typically around the 
                    thickness of the filament. This and the tile width determines the resampled resolution of the DEM 
                    raster that is the basis of the mesh. E.g. if you want your tile to be 100 mm wide and your printres 
                    is set to 0.25 mm, the DEM raster will be re-sampled from its original resolution to 
                    the equivalent of 400 cells.
                    If the tile is 4000 m wide in reality, each cell would cover 10 m, about the resolution of NED.
                    It is probably pointless to ask for a resolution below the original DEM by lowering printres
                    to less than 0.25 in this case.
                    To NOT resample the DEM but instead use the original resolution of your raster, set printres to -1 
                  
 * `tile_centered`: 
   * True:  each tile is centered around 0/0
   * False: all tiles are offset so the all "fit together" when they all are loaded into a 3D viewer, such as Meshlab

 * `zip_file_name`: Prefix of output filename. The end is the datetime of the
                    file's creation.

 * `zscale`:        Vertical exaggeration versus horizontal units.
 
 * `CPU_cores_to_use`: Number of CPU cores (processes) to use. 0 means: use all available cores, null does not use multiprocessing, which is useful when running the code in a Debugger.
 
 * `max_cells_for_memory_only`: if the nuymber of raster cells is bigger than this number (default: 1000 * 1000), temp files are used in the later stages of processing. This is slower but less memory intensive than assembling the entire zip file in memory alone.
 
 * `importedDEM`: default: null . With the default, the DEM is fetched from Google Earth Engine as detailed above. If it is set to a filename, such as `pyramid.tif`, this file is used as DEM. In this case DEM_name, bllat, bllon, trlat and trlon are ignored, but all other parameters are still used. You can test this with pyramid.tif (in standalone folder) which has an elevation of 0 to 255, so probably will need a z-scale of 0.5 on a width of 100 mm.  Note that this requires gdal (https://trac.osgeo.org/gdal/wiki/GdalOgrInPython) to be installed. Any GDAL raster format (http://www.gdal.org/frmt_various.html) should be readable.

A note on distances: Google Earth Engine requires that the requested area is given in lat/lon coordinates but it's worth knowing the approximate real-world meter distance in order to select good values for the tile width, number of tiles and the printres. The server version displays the tile width in Javascript but for the standalone version you need to calculate it yourself. This haversine distance (https://en.wikipedia.org/wiki/Haversine_formula, interactive calulator here: http://www.movable-type.co.uk/scripts/latlong.html) depends on the latitude of your area. Once you know the width of your tile in meters, divide it by the number of cells along x (400 cells in the example above) to get an idea of the re-sampled real-world resolution of your model and its scale. The server Help file (https://docs.google.com/document/d/1GlggZ47xER9N85Qls_MiE1jNuihlYEZnFFSVZtX8bKU/pub) goes into the interplay of these parameters in the section: Understanding the linkage of tile size, tile number, source DEM resolution and 3D print resolution




Server
======

Running `TouchTerrain_app.py` starts a server module (service) which will be run inside Apache. Modify the ansiable script to set up the server. The server creates a webpage, through which the user inputs the area selection and print parameters.

The server requires that the file `config.py` be edited to appropriate values.The server also requires oauth authentication to run with Earth Engine. You will need to obtain a private key (`.pem`) file and edit `config.py` to point to it.

The server presents users with `index.html`, which can be styled to suit your needs, provided the various input dialogs and JavaScript remain.

index.html contain the standard Google Analytics tracking boilerplate code at the top. By default it has a bogus tracking id ('UA-XXXXXXXX'), so tracking doesn't work. If you want to activate it, you need to set up your own Google Analytics (https://www.google.com/analytics/web/#home) and use your own UA- tracking code instead.

touchterrain_config.py contains server specific config settings:
- NUM_CORES:  0 means: use all cores, 1 means: use only 1 core (usefull for debugging)
- MAX_CELLS: if the raster has more cells than this number, tempfiles are used instead of memory during the later stages of processing. This is slower but less memory intensive than assembling the entire zip file in memory alone.
- MAX_LOAD_FACTOR: if width / printres * num tiles is bigger than this number, processing is not started. This prevents jobs that are so big that the server runs out of memory. It is, however, just a heuristic. Recommeded practice is to start a job and see if virtual memory (swapspace) is used and to lower MAX_LOAD_FACTOR until this does not happen. 


Common
======

The `common` directory contains files used by both the standalone and server apps.


NEWS
Sep. 25:
added optional Google Maps key. If a file called GoogleMapsKey.txt in the serve folder contains a valid google maps key, it will be inlined into index.html which should result in a proper Google Map. Not having this file will result in a  Google Map that's crappy but still works. Having a invalid key will result in no map at all.

Sept. 18:
- fixed bug where the print resolution was wrong in some cases when multiple tiles are used
- added the geotiff I get from Google Earth Engine to the zipped folder as DEM.tif 

August 29:
- both: set resample mode for GEE to bilinear (from nearest neighbor), which was creating aliasing artifacts when projecting/resampling. Standalone resampling is still nearest neighbor.
- standalone: printers = -1 for using the original raster file resolution


April 4, 2018:
- server: added MAX_LOAD factor which prevents running of large jobs, that would requite too much memory.
- standalone: added max_cells_for_memory_only for use of tempfiles instead of in-memory for processing large rasters.

Feb. 14, 2018: 
- added the use of tempfiles instead of in-memory for processing large rasters 
- config settings/constants are now in common/touchterrain_config.py

Jan. 12, 2018: 
- added code for zigzag magic (currently disabled). May help to prevent curling corners.
- vertex indexing is now only done for obj, not for STL

Dec. 05, 2017: 
- Cleaned up index.html. 
- Added getting area box from kml file. 
- All GUI options are now put in URL. 
- Changed tile width/height from cm to mm. 
- Gamma value changes now trigger refresh.

Oct. 27, 2017: server: fixed Earth engine 404 bug. Standalone: added ability to read in local geotiff file (via json config file). Added link to blog. 
Aug. 14, 2017: added a show URL button to the Area UI. This will load the app (w/o changing the area!) and show the full URL in the browser. This is useful to capture the area for others to use.

Aug. 11, 2017: added multi-core processing. By default server and standalone will use all available cores. If this doesn;t work, CPU_cores_to_use can be set to 1 to go back to single-processor mode

Apr. 18, 2017: added more options to the pulldown lists for tile size and z-scale
Apr. 4: added check to convert water cells outside the defines DEM area to 0. Does not affect onshore cells with <0 elevation (e.g. holland)

Mar. 6: index.html
- added boilerplate code for google analytics. Change UA-XXXXXX to your tracking id if you want to use it
- added contact info and github URL
 
Feb. 16: separated standalone and server version into subfolders:
- common: .py files (modules) used by both versions
- standalone: main for standalone, install-howto, etc.. Works
- server: server main (touchterrain_app.py) and server setup files. 

version 0.12 is now live at: touchterrain.geol.iastate.edu 

Jan. 23: changed the order in which ee is intialized. This enables the stand alone version to use any google account for authentication, rather than having to go through config.py (server still uses config.py)

April 3:  added check to set all <0 elevations to 0 for non-bathymetry DEM sources (NED, STRM). That 

0.12 (Dec. 12 - X-mas edition): 
- Transparency defaults to 40%
- results page has link to viewstl.com to preview downloaded files
- ETOPO elevation is for ice, not bedrock
- can now use server type flags: Apache, paste or GEA_devserver
- added a hillshade gamma slider to change default gamma (1.0). Requires reload of page: press set new gamma button.
- added 30m SRTM DEM as data source
- better info on how the original raster DEM will be rescaled based on area, tile number/size, etc. 
- tile height (in cm) is now automatically calculated from tile width and aspect ratio of red box (using meters for box sides).
- general layout changes, incl. a link to help page.

0.11 (Oct. 31 - Halloween edition): Includes Levi's changes/fixes/additions:
- proper separation of server vs stand-alone:
- stand-alone version can now be run without having to install the google app engine (GAE) modules
- full server now runs on Apache mod_wsgi, rather than via the Google App Engine dev server. However, the GAE modules still need to be installed, as some parts are used by the server module.
- fixed a noobish use of a Python global for duplicating request_handling data for the pre-flight page


Version 0.10 of the TouchTerrain project, primarily a set of python source code files
for Python 2.7:

- TouchTerrain-app.py: a server module (service) to be run as part of a Google App Engine server. The server creates a webpage, through which the user inputs the area selection and print parameters.
    
- index.html: HTML template for the main webpage, includes Javascript
    
- TouchTerrain_standalone.py: A stand-alone version in which the user input is given 
    in a JSON file, rather then via a web page.
    
- TouchTerrainEarthEngine.py: With the user input, gets the DEM raster (geotiff) from the Google Earth Engine data server and, using the grid class, creates the 3D models (tiles).
    
- grid_tesselate.py: defines the grid class used to create a triangle "mesh" and save it in the desired file format (STL or OBJ)
    
- Coordinate_system_conv.py, InMemoryZip.py: utility functions  

- config.py: used for oauth credentials for the Google dev (Earth Engine) account 

- tmp folder: contains an example terrain model, a zipped stl file 



TouchTerrain_app.py
"""TouchTerrain-app - a server module"""

'''
@author:     Chris Harding
@license:    GPL
@contact:    charding@iastate.edu

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''


# CH 11/21/2016: Added support for hillshade gamma. Unlike hs opacity, it needs to be set on server side 
# and so requires a reload
# CH 11/16/2016: added flags for server will run: Apache, GAE_devserver, or paste
# CH 06/20/2016: added ETOPO and changed the DEM source strings, see DEM_sources from TouchTerrainEarthEngine.py
# CH 12/01/2016: MainPage() request handler now uses GET, the querie string contains the coords of the window, the DEM and the coords of the region
# CH 12/22/2015: changes UI for switching (no need to click button) and clipped SRTM data to 0 for offshore (was -32768)
# CH 12/16/2015: added switch between 10m (NED) and 90m (SRTM) DEM
# CH 12/01/2015: created a way to remotely disable the write restrictions on the devserver via a fake __init__
# CH 11/23/2015: went back to no-threading as it didn't seem to work in the devserver
#    added a pre-flight page to warn the user of no feedback. Had to fake the request part via a global var (eeewwww!))

import math
import os
#import threading
from datetime import datetime
import ee      # can be installed under site-packges


SERVER_TYPE = "Apache" # "paste" or "GAE_devserver" or "Apache"
#SERVER_TYPE = "paste" # so I can run the server inside a debugger ...

#  set sys.path to include all modules in google_appengine\lib
if SERVER_TYPE == "GAE_devserver":
    import sys, glob
    GA_lib = r"C:\Program Files (x86)\Google\google_appengine\lib" # location of google_appengine\lib is on your system
    for d in glob.glob(GA_lib + os.sep + "*"):
        sys.path.append(d) # append each folder to the sys.path, which is searched on import
    sys.path.append(r"C:\Program Files (x86)\Google\google_appengine") 

# These modules require that the path to GAE has been set, either via above or via using the GAE laucher
import TouchTerrainEarthEngine # my modules
import config  # config.py must be in this folder, use GAE modules
import webapp2
import jinja2
jinja_environment = jinja2.Environment(loader=jinja2.FileSystemLoader(os.path.dirname(__file__)))
import logging
import time
logging.Formatter.converter = time.gmtime # set to gmt as GAE uses that,

tmp_folder = "/var/www/html/touchterrain/tmp"  # dir to store zip files on the dev server

# functions for computing hillshades in EE (from EE examples)
# Currently, I only use the precomputed hs, but they might come handy later ...
def Radians(img):
  return img.toFloat().multiply(math.pi).divide(180)

def Hillshade(az, ze, slope, aspect):
  """Compute hillshade for the given illumination az, el."""
  azimuth = Radians(ee.Image(az))
  zenith = Radians(ee.Image(ze))
  # Hillshade = cos(Azimuth - Aspect) * sin(Slope) * sin(Zenith) +
  #     cos(Zenith) * cos(Slope)
  return (azimuth.subtract(aspect).cos()
          .multiply(slope.sin())
          .multiply(zenith.sin())
          .add(
              zenith.cos().multiply(slope.cos())))



#
# The page for selecting the ROI and putting in printer parameters
#

# example query string: ?DEM_name=USGS%2FNED&map_lat=44.59982&map_lon=-108.11694999999997&map_zoom=11&trlat=44.69741706507476&trlon=-107.97962089843747&bllat=44.50185267072875&bllon=-108.25427910156247&hs_gamma=1.0
class MainPage(webapp2.RequestHandler):
  def get(self):                             # pylint: disable=g-bad-name

    ee.Initialize(config.EE_CREDENTIALS, config.EE_URL)
    #print self.request.GET # all args

    DEM_name = self.request.get("DEM_name")
    map_lat = self.request.get("map_lat")
    map_lon = self.request.get("map_lon")
    map_zoom = self.request.get("map_zoom")
    trlat = self.request.get("trlat")
    trlon = self.request.get("trlon")
    bllat = self.request.get("bllat")
    bllon = self.request.get("bllon")
    hillshade_gamma = self.request.get("hs_gamma")

    # default args:
    if DEM_name not in TouchTerrainEarthEngine.DEM_sources: DEM_name = 'USGS/NED' # 10m NED as default
    if map_lat == "": map_lat = "44.59982"  # Sheep Mtn, Greybull, WY
    if map_lon == "": map_lon ="-108.11695"
    if map_zoom == "": map_zoom ="11"
    if hillshade_gamma == "": hillshade_gamma = "1.0"


    # for ETOPO1 we need to first select one of the two bands as elevation
    if DEM_name == """NOAA/NGDC/ETOPO1""":
	img = ee.Image(DEM_name)
	elev = img.select('bedrock') # or ice_surface
	terrain = ee.Terrain.products(elev)
    else:
	terrain = ee.Algorithms.Terrain(ee.Image(DEM_name))

    hs = terrain.select('hillshade')

    mapid = hs.getMapId( {'gamma':float(hillshade_gamma)}) # opacity is set in JS

    # jinja will inline these variables and their values into the template and create index.html
    template_values = {
        'mapid': mapid['mapid'],
        'token': mapid['token'],
        'DEM_name': DEM_name,
        'map_lat': map_lat,
        'map_lon': map_lon,
        'map_zoom': map_zoom,
        'map_zoom': map_zoom,
        'trlat' : trlat,
        'trlon' : trlon,
        'bllat' : bllat,
        'bllon' : bllon,
        "hsgamma": hillshade_gamma,
    }

    # this creates a index.html "file" with mapid, token, etc. inlined
    template = jinja_environment.get_template('index.html')
    self.response.out.write(template.render(template_values))

    # delete files in tmp that are >24 hrs old is done by a cron shell script rather then by the dev server

if SERVER_TYPE == "GAE_devserver":    
    # fake fake file init - this allows me to instantiate a FakeFile object with write-only checks disabled.
    # Note that I still can't use any other OS functions on files, e.g. get their time or delete them!
    # To delete old files, I use a chron job.
    from google.appengine.tools.devappserver2.python import stubs
    def fake__init__(self, filename, mode='r', bufsize=-1, **kwargs):
		"""Initializer. See file built-in documentation."""
    
		'''# CH: disabled checks
		if mode not in FakeFile.ALLOWED_MODES:
		  raise IOError(errno.EROFS, 'Read-only file system', filename)
    
		if not FakeFile.is_file_accessible(filename):
		  raise IOError(errno.EACCES, 'file not accessible', filename)
		'''
		super(stubs.FakeFile, self).__init__(filename, mode, bufsize, **kwargs)
    
    stubs.FakeFile.__init__ = fake__init__ # now overwrite __init__ with my version that doesn't disable writing like the official version

# preflight page: showing some notes on how there's no used feedback until processing is done
# as I don't know how to carry over the args I get here to the export page handler's post() method,
# I store the entire request as a global and write it back later. TODO: find a better way!
class preflight(webapp2.RequestHandler):
    def __init__(self, request, response):
    	# Set self.request, self.response and self.app.
        self.initialize(request, response)
        app = webapp2.get_app()
        app.registry['preflightrequest'] = self.request
        print app.registry['preflightrequest'] #Levi I don't know why, but without this, complains about expired request

    def post(self):
	#print self.request.POST # all args
	self.response.headers['X-Content-Type-Options'] = 'nosniff'	# Prevent browsers from MIME-sniffing the content-type:
	self.response.headers["X-Frame-Options"] = "SAMEORIGIN"   # prevent clickjacking
	self.response.out.write('<html><body>')
	self.response.out.write("<h2>Processing started:</h2>")
	self.response.out.write("Press the Start button to process the DEM into 3D model files.<br>")
	self.response.out.write("Note that there's NO progress indicator (yet), you will only see this page trying to connect. That's OK, just be patient!<br>")
	self.response.out.write("Pressing Start again during processing has no effect.<br>")
	self.response.out.write("Once your 3D model is created, you will get a new page (Processing finished) for downloading them in a zip file.<br><br>")
	self.response.out.write('<form action="/export" method="POST" enctype="multipart/form-data">')
	self.response.out.write('<input type="hidden" maxlength="50" size="20" name="Note" id="Note" value="NULL">') # for user comment
	#self.response.out.write('<input type="hidden" name="prog_pct" id="prog_pct" value="0">') # progress percentage
	self.response.out.write('<input type="submit" value="Start"> </form>')
	self.response.out.write('</body></html>')

# Page that creates the 3D models (tiles) in a in-memory zip file, stores it in tmp with
# a timestamp and shows a download URL to the zip file. The args are the same as in the
# main page (via preflight by using the global) The write to tmp dir is only possible
# b/c I disable the write-only checks temporarily. This folder must already exist as it
# cannot be created by the devserver.
class ExportToFile(webapp2.RequestHandler):
    def __init__(self, request, response):
        self.initialize(request, response)
        app = webapp2.get_app()
        self.request = app.registry.get('preflightrequest')

    def post(self): # make tiles in zip file and write
	#print self.request.arguments() # should be the same as given to preflight
	self.response.headers['X-Content-Type-Options'] = 'nosniff'	# Prevent browsers from MIME-sniffing the content-type:
	self.response.headers["X-Frame-Options"] = "SAMEORIGIN"   # prevent clickjacking
	self.response.out.write('<html><body>')
	self.response.out.write('<h2>Processing finished:</h2>')

	# debug: print/log all args and then values
	args = {} # put arg name and value in a dict as key:value
	for k in ("DEM_name", "trlat", "trlon", "bllat", "bllon", "printres", "ntilesx", "ntilesy", "tilewidth",
	          "basethick", "zscale", "fileformat"):
	    v = self.request.get(k) # key = name of arg
	    args[k] = v # value
	    if k not in ["DEM_name", "fileformat"]: args[k] = float(args[k]) # floatify non-string args
	    if k[:4] == "tile": args[k] = args[k] * 10 # multiply tilewidth/height by 10 to get from cm to mm
	    #print k, args[k]
	    self.response.out.write("%s = %s <br>" % (k, str(args[k])))
	    logging.info("%s = %s" % (k, str(args[k])))

	# name of file is seconds since 2000
	myname = str(int((datetime.now()-datetime(2000,1,1)).total_seconds() * 1000))

        # create zip and write to tmp
	str_buf = TouchTerrainEarthEngine.get_zipped_tiles(**args) # all args are in a dict
	fname = myname + ".zip" # create filename for zipped
	logging.info("About to write: " + tmp_folder + os.sep + fname)
	fname = tmp_folder + os.sep + fname # put in tmp folder
	if SERVER_TYPE == "GAE_devserver":
	    f = stubs.FakeFile(fname, "wb+") # make a fake Fakefile instance (with devserver's write restriction disabled)
	else:
	    f = open(fname, "wb+") # write to folder
	f.write(str_buf)
	f.close()
	logging.info("finished writing %s" % (myname))

	#str_buf = TouchTerrain.get_zipped_tiles(**args)
	#str_buf = TouchTerrain.get_zipped_tiles("USGS/NED", ntilesx=2, ntilesy=2, **args)
	#self.response.headers['Content-Type'] = 'text/zip'
	#self.response.write(str_buf)

	self.response.out.write('<br><form action="tmp/%s.zip" method="GET" enctype="multipart/form-data">' % (myname))
	self.response.out.write('<input type="submit" value="Download zip File " title="">   (will be deleted in 24 hrs)</form>')
	#self.response.out.write('<form action="/" method="GET" enctype="multipart/form-data">')
	#self.response.out.write('<input type="submit" value="Go back to selection map"> </form>')
	self.response.out.write("<br>To return to the selection map, click the back button in your browser twice")
	self.response.out.write(
	"""<br>After downloading you can preview a STL/OBJ file at <a href="http://www.viewstl.com/" target="_blank"> www.viewstl.com ) </a>  (limit: 35 Mb)""")

# the pages that can be requested from the browser and the handler that will respond (get or post method)
app = webapp2.WSGIApplication([('/', MainPage), # index.html
                              ('/export', ExportToFile), # results page, generated by: <form action="/export" ....>
                              ('/preflight', preflight)],
                              debug=True)

if SERVER_TYPE == "GAE_devserver":
    # Running this as a Google App Engine module via its development app server (cloud.google.com/appengine/docs/python/download)
    # I assume that you have the python App Engine installed in something like google_appengine and that
    # file, the other python files, the app,yaml file, tmp and doc folders are all in a folder (say touchterrain) that's inside google_appengine.
    # At the beginning of the file you've added path to sys.path so it finds the GAE modules needed.
    # the google_appengine folder should comtain dev_appserver.py, open a terminal, go into google_appengine and run:
    # python dev_appserver.py --host myserver.whatever.edu touchterrain
    # Alternatively, you can use the GAE app launcher
    pass


if SERVER_TYPE == "paste":
   # You only need the following if you want to run your app directly (instead of the launcher/devserver),
   # so you can debug it directly (see http://webapp-improved.appspot.com/tutorials/quickstart.nogae.html for more)
   # You will need to install the Paste module and hack the path to google appengine modules (see above)    
    from paste import httpserver 
    print "running local httpserver ,", 
    httpserver.serve(app, host='127.0.0.1', port='8080') # run the server
print "done"
